src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+#ifdef _WIN32
+#ifndef FD_SETSIZE
+#define FD_SETSIZE 2048  // Set max fds - only for select() fallback
+#endif
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include <atomic>
+#include "tinycthread.h"
+#include "later.h"
+#include "callback_registry_table.h"
+
+#if R_VERSION < R_Version(4, 5, 0)
+inline SEXP R_mkClosure(SEXP formals, SEXP body, SEXP env) {
+  SEXP fun = Rf_allocSExp(CLOSXP);
+  SET_FORMALS(fun, formals);
+  SET_BODY(fun, body);
+  SET_CLOENV(fun, env);
+  return fun;
+}
+#endif
+
+extern CallbackRegistryTable callbackRegistryTable;
+extern SEXP later_fdcancel;
+extern SEXP later_invisibleSymbol;
+
+#define LATER_INTERVAL 1024
+
+#ifdef _WIN32
+#define POLL_FUNC WSAPoll
+#else
+#define POLL_FUNC poll
+#endif
+
+typedef struct ThreadArgs_s {
+  std::shared_ptr<std::atomic<bool>> flag;
+  std::unique_ptr<std::vector<int>> fds;
+  SEXP callback;
+  double timeout;
+  int rfds, wfds, efds;
+  int num_fds;
+  int loop;
+} ThreadArgs;
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  const bool flag = args->flag->load();
+  args->flag->store(true);
+  if (!flag) {
+    std::memcpy((void *) DATAPTR_RO(CADR(args->callback)), args->fds->data(), args->num_fds * sizeof(int));
+    Rf_eval(args->callback, R_GlobalEnv);
+  }
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: if necessary to add method for HANDLES on Windows. Would be different code to SOCKETs.
+// TODO: implement re-usable background thread.
+static int wait_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  int *fds = args->fds->data();
+
+  int result;
+
+  const bool infinite = args->timeout == R_PosInf;
+  int timeoutms = infinite || args->timeout < 0 ? 1000 : (int) (args->timeout * 1000);
+  int repeats = infinite || args->timeout < 0 ? 0 : timeoutms / LATER_INTERVAL;
```

Comment:
I find this confusing and also I think it assumes that this thread will be promptly scheduled when it wants to wake up. I've always handled this situation (multiple little waits with an overall timeout) by adding timeoutms to the current time at the start of the overall operation, and then for each little wait, wait for the lesser of (time til expiration) and (liveness interval). For example, https://github.com/r-lib/later/blob/968542a0b42fa191e6a5a57f180a6af345083bf5/src/callback_registry.cpp#L373-L402

(Note that the Timestamp class used there is designed specifically for this kind of waiting, it won't be affected by the system clock being messed with)

Reply:
Yes, apologies here - I had meant to write some comments for this logic!

The advantage of my method is that it just relies on arithmetic rather than calls for the system time after each loop. The disadvantage is that there isn't the one source of truth being the fixed expiry time. The slippage on the wake-up time is taken as a given as the semantics for a timeout is typically "wait for at least x ms", without the expectation for too fine a granularity.

However, I do like how the Timestamp class is implemented. The intent is at least much clearer so I've proceeded to adopt it in [d8eb35f](https://github.com/r-lib/later/pull/190/commits/d8eb35f7f564de5fe84e78b2a17c0929cd91c139) (and it also makes sense as it's already in the package), with the following notes:

1. Made sure it runs at least once even when the timeout is set as 0.
2. Poll interval is set to a default of 1024 ms, which means that for normal use in curl the default of 1000ms won't be broken up by this.
3. Initialized the timestamp as soon as possible up front instead of on the thread.